jvm/src/main/java/com/microsoft/bond/protocol/CompactBinaryProtocolWriter.java

```diff
@@ -0,0 +1,383 @@
+
+    public CompactBinaryProtocolWriter(OutputStream outputStream, short version) {
+        this.writer = new BinaryStreamWriter(outputStream);
+        this.version = version;
+        this.firstPassCounter = (version == 2) ? new FirstPassCounter() : null;
+        this.currentStructIndex = 0;
+    }
+
+    @Override
+    public void writeVersion() throws IOException {
+        this.writer.writeInt16(MAGIC);
+        this.writer.writeInt16(this.version);
+    }
+
+    @Override
+    public void writeStructBegin(Metadata metadata) throws IOException {
+        if (this.version == 2) {
+            int structLength = this.firstPassCounter.getResultLength(this.currentStructIndex);
+            this.writer.writeVarUInt32(structLength);
+        }
+
+        ++this.currentStructIndex;
+    }
+
+    @Override
+    public void writeBaseBegin(Metadata metadata) throws IOException {}
+
+    @Override
+    public void writeStructEnd() throws IOException {
+        this.writer.writeInt8((byte)BondDataType.BT_STOP.value);
+
+        --this.currentStructIndex;
+    }
+
+    @Override
+    public void writeBaseEnd() throws IOException {
+        this.writer.writeInt8((byte)BondDataType.BT_STOP_BASE.value);
+    }
+
+    @Override
+    public void writeFieldBegin(BondDataType type, int id, Metadata metadata) throws IOException {
+        if (id <= 5) {
+            this.writer.writeInt8((byte)(type.value | (id << 5)));
+        }
+        else if (id <= 0xFF) {
+            this.writer.writeInt8((byte)(type.value | 0xC0));
+            this.writer.writeInt8((byte)id);
+        }
+        else {
+            this.writer.writeInt8((byte)(type.value | 0xE0));
+            this.writer.writeInt16((short)id);
+        }
+    }
+
+    @Override
+    public void writeFieldEnd() throws IOException {}
+
+    @Override
+    public void writeFieldOmitted(BondDataType type, int id, Metadata metadata) throws IOException {}
+
+    @Override
+    public void writeContainerBegin(int count, BondDataType elementType) throws IOException {
+        if (this.version == 2 && count < 7) {
+            this.writer.writeInt8((byte)(elementType.value | ((count + 1) << 5)));
+        }
+        else {
+            this.writer.writeInt8((byte)elementType.value);
+            this.writer.writeVarUInt32(count);
+        }
+    }
+
+    @Override
+    public void writeContainerBegin(int count, BondDataType keyType, BondDataType valueType) throws IOException {
+        this.writer.writeInt8((byte)keyType.value);
+        this.writer.writeInt8((byte)valueType.value);
+        this.writer.writeVarUInt32(count);
+    }
+
+    @Override
+    public void writeContainerEnd() throws IOException {}
+
+    @Override
+    public void writeInt8(byte value) throws IOException {
+        this.writer.writeInt8(value);
+    }
+
+    @Override
+    public void writeInt16(short value) throws IOException {
+        this.writer.writeVarUInt16(ZigzagHelper.encodeZigzag16(value));
```

Comment:
I think your helpers are decent candidates for `import static`. Qualifying with `ZigzagHelper.` burns enough keystrokes for little enough clarity gain that I'd lean that way.